dnf/cli/option_parser.py

```diff
@@ -377,38 +366,45 @@ def get_usage(self):
                     usage += "%-25s %s\n" % (name, summary)
         return usage
 
-    def add_argument(self, *args, **kwargs):
-        # process options and arguments separately
+    def _add_command_options(self, command):
+        self.prog = "%s %s" % (self.prog, command._basecmd)
+        self.description = command.summary
+        self.command_positional_parser = argparse.ArgumentParser(self.prog, add_help=False)
+        self.command_positional_parser.print_usage = self.print_usage
+        self.command_positional_parser._positionals.title = None
+        self.command_group = self.add_argument_group(
+            'Command "{}" specific options'.format(command._basecmd))
+        self.command_group.add_argument = self.cmd_add_argument
+        command.set_argparser(self.command_group)
+
+    def cmd_add_argument(self, *args, **kwargs):
         if all([(arg[0] in self.prefix_chars) for arg in args]):
-            return super(OptionParser, self).add_argument(*args, **kwargs)
+            return argparse._ArgumentGroup.add_argument(self.command_group, *args, **kwargs)
         else:
-            return self.command_arg_parser.add_argument(*args, **kwargs)
+            return self.command_positional_parser.add_argument(*args, **kwargs)
 
     def parse_main_args(self, args):
-        parser = argparse.ArgumentParser(add_help=False,
-                                         parents=[self.main_parser])
-        parser.add_argument('command', nargs='?', help=argparse.SUPPRESS)
-        namespace, _unused_args = parser.parse_known_args(args)
+        namespace, _unused_args = self.parse_known_args(args)
         return namespace
 
     def parse_command_args(self, command, args):
-        self._command_parser(command)
-        namespace, extras = super(OptionParser,
-                                  self).parse_known_args(args)
-        command.opts = self.command_arg_parser.parse_args(extras, namespace)
+        self._add_command_options(command)
+        namespace, unused_args = self.parse_known_args(args)
+        namespace = self.command_positional_parser.parse_args(unused_args, namespace)
+        command.opts = namespace
         return command.opts
 
     def print_usage(self, file_=None):
-        # pylint: disable=E0202,W0212
-        self._actions += self.command_arg_parser._actions
+        if self.command_positional_parser:
+            self._actions += self.command_positional_parser._actions
         super(OptionParser, self).print_usage(file_)
 
     def print_help(self, command=None):
         # pylint: disable=W0212
         if command:
-            cp = self._command_parser(command)
-            cp._actions += cp.command_arg_parser._actions
-            super(OptionParser, cp).print_help()
+            self._add_command_options(command)
+            self._actions += self.command_positional_parser._actions
+            self._action_groups.append(self.command_positional_parser._positionals)
```

Comment:
Also private?

Reply:
These private members were actually used even before the patch.

Reply:
But these are hard ones. Removing them would mean that the whole option parsing part should be rewritten (not saying it doesn't need to be). But the primary goal of this pull request was to make help for command clearer. Given that these private members (well, not all of them) were already used before this patch, do you think it is possible to merge this patch?